nsqd/nsqd.go

```diff
@@ -239,40 +231,46 @@ func (n *NSQD) RemoveClient(clientID int64) {
 	n.clientLock.Unlock()
 }
 
-func (n *NSQD) Main() {
+func (n *NSQD) Main() error {
 	var err error
 	ctx := &context{n}
 
 	n.tcpListener, err = net.Listen("tcp", n.getOpts().TCPAddress)
 	if err != nil {
-		n.logf(LOG_FATAL, "listen (%s) failed - %s", n.getOpts().TCPAddress, err)
-		os.Exit(1)
+		return fmt.Errorf("listen (%s) failed - %s", n.getOpts().TCPAddress, err)
 	}
 	n.httpListener, err = net.Listen("tcp", n.getOpts().HTTPAddress)
 	if err != nil {
-		n.logf(LOG_FATAL, "listen (%s) failed - %s", n.getOpts().HTTPAddress, err)
-		os.Exit(1)
+		return fmt.Errorf("listen (%s) failed - %s", n.getOpts().HTTPAddress, err)
 	}
 	if n.tlsConfig != nil && n.getOpts().HTTPSAddress != "" {
 		n.httpsListener, err = tls.Listen("tcp", n.getOpts().HTTPSAddress, n.tlsConfig)
 		if err != nil {
-			n.logf(LOG_FATAL, "listen (%s) failed - %s", n.getOpts().HTTPSAddress, err)
-			os.Exit(1)
+			return fmt.Errorf("listen (%s) failed - %s", n.getOpts().HTTPSAddress, err)
 		}
 	}
 
+	exitCh := make(chan error)
+	var once sync.Once
+	exitFunc := func(err error) {
+		once.Do(func() {
+			n.logf(LOG_FATAL, "%s", err)
```

Comment:
I think this will still happen once during normal shutdown, usually with a `nil` error